File: graylog2-rest-client/src/main/java/org/graylog2/restclient/models/Startpage.java
/**
 * This file is part of Graylog2.
 *
 * Graylog2 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Graylog2 is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Graylog2.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.graylog2.restclient.models;

/**
 * @author Lennart Koopmann <lennart@torch.sh>
 */
public class Startpage {

    public enum Type {
        STREAM("Stream"),
        DASHBOARD("Dashboard");

        private final String description;

        Type(String description) {
            this.description = description;
        }

        public String getDescription() {
            return description;
        }
    }

    private final Type type;
    private final String id;

    public Startpage(Type type, String id) {
        this.type = type;
        this.id = id;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Startpage startpage = (Startpage) o;

        if (id != null ? !id.equals(startpage.id) : startpage.id != null) return false;
        if (type != startpage.type) return false;

        return true;
    }

    @Override
    public int hashCode() {
        int result = type != null ? type.hashCode() : 0;
        result = 31 * result + (id != null ? id.hashCode() : 0);
        return result;
    }

    public Type getType() {
        return type;
    }

    public String getId() {
        return id;
    }

}

File: graylog2-web-interface/src/api/startpage.ts
/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import ApiRoutes from 'routing/ApiRoutes';
import UserNotification from 'util/UserNotification';
import { qualifyUrl } from 'util/URLUtils';
import fetch from 'logic/rest/FetchProvider';
import { CurrentUserStore } from 'stores/users/CurrentUserStore';

export const setStartpage = (userId: string, type: string, id: string): Promise<void> => {
  const url = qualifyUrl(ApiRoutes.UsersApiController.update(userId).url);
  const payload: { type?: string; id?: string } = {};

  if (type && id) {
    payload.type = type;
    payload.id = id;
  }

  return fetch('PUT', url, { startpage: payload }).then(
    () => {
      CurrentUserStore.reload();
      UserNotification.success('Your start page was changed successfully');
    },
    (error: unknown) =>
      UserNotification.error(
        `Changing your start page failed with error: ${error}`,
        'Could not change your start page',
      ),
  );
};
